dvc/remote/local/__init__.py

```diff
@@ -573,3 +573,53 @@ def _get_unpacked_dir_names(self, checksums):
             if self.is_dir_checksum(c):
                 unpacked.add(c + self.UNPACKED_DIR_SUFFIX)
         return unpacked
+
+    def _needs_checkout(self, path_info, checksum):
+        # NOTE: In case if path_info is already cached, and cache type is
+        # 'copy' we would like to avoid relinking.
+        # `symlink/hardlink/reflink` is cheap, but if we need `copy` links,
+        # then to save on copying we should only relink if we already have
+        # `hardlink` or a `symlink`, we don't care about reflinks, because
+        # they are indistinguishable from copy anyway.
+
+        if (
+            self._is_cache_copy(path_info)
+            and not System.is_hardlink(path_info)
+            and not System.is_symlink(path_info)
+            and not self.changed(path_info, {self.PARAM_CHECKSUM: checksum})
+        ):
+            if self.protected:
+                self.protect(path_info)
+            else:
+                # NOTE: we can unprotect, because `hardlink/symlink` check
+                # has been performed before, so no chance of copying,
+                # only chmod-ing.
+                self.unprotect(path_info)
+
+            msg = "File '{}' didn't change"
+            logger.debug(msg.format(str(path_info)))
+            return False
+        return True
+
+    def _is_cache_copy(self, path_info):
+        # NOTE: path_info required to make test reliable, when cache is on
+        # different fs than path_info
+
+        test_cache_file = self.path_info / ".cache_type_test_file"
+
+        if self.cache_types[0] == "copy":
+            self.remove(test_cache_file)
+            return True
+
+        workspace_file = path_info.parent / "." + uuid()
+
+        if not self.exists(test_cache_file):
+            with open(fspath_py35(test_cache_file), "wb") as fobj:
```

Comment:
```python
test_cache_file.write_bytes(bytes(1))
```

Reply:
done

Reply:
reversed, wont fix, PosixPathInfo descends from PurePath which does not have this method (Path does).